docs/custom-transfers.md

```diff
@@ -0,0 +1,276 @@
+
+Or if there was an error in the transfer:
+
+```json
+{ "id":"complete", "oid": "bf3e3e2af9366a3b704ae0c31de5afa64193ebabffde2091936ad2e7510bc03a", "error": { "code": 2, "message": "Explain what happened to this transfer" }}
+```
+
+* `id`: Always "complete" to identify this message
+* `oid`: the identifier of the LFS object
+* `error`: Should contain a `code` and `message` explaining the error
+
+##### Downloads
+
+For downloads the request sent from git-lfs to the transfer process will look
+like this:
+
+```json
+{ "id":"download", "oid": "22ab5f63670800cc7be06dbed816012b0dc411e774754c7579467d2536a9cf3e", "size": 21245, "action": { "href": "nfs://server/path", "header": { "key": "value" } } }
+```
+
+* `id`: Always "download" to identify this message
+* `oid`: the identifier of the LFS object
+* `size`: the size of the LFS object
+* `action`: the "download" action copied from the response from the batch API.
+  This contains "href" and "header" contents, which are named per HTTP
+  conventions, but can be interpreted however the custom transfer agent wishes
+  (this is an NFS example, but it doesn't even have to be an URL). Generally,
+  "href" will give the primary connection details, with "header" containing any
+  miscellaneous information needed.
+
+Note there is no file path included in the download request; the transfer 
+process should create a file itself and return the path in the final response
+after completion (see below).
+
+The transfer process should post one or more [progress messages](#progress) and 
+then a final completion message as follows:
+
+```json
+{ "id":"complete", "oid": "22ab5f63670800cc7be06dbed816012b0dc411e774754c7579467d2536a9cf3e", "path": "/path/to/file.png", "error": null}
+```
+
+* `id`: Always "complete" to identify this message
+* `oid`: the identifier of the LFS object
+* `path`: the path to a file containing the downloaded data, which the transfer
+  process reliquishes control of to git-lfs. git-lfs will move the file into LFS
+  storage.
+* `error`: Should be null if there was no error
+
+Or, if there was a failure transferring this item:
+
+```json
+{ "id":"complete", "oid": "22ab5f63670800cc7be06dbed816012b0dc411e774754c7579467d2536a9cf3e", "error": { "code": 2, "message": "Explain what happened to this transfer" }}
+```
+
+Errors for a single transfer request should not terminate the process. The error
+should be returned in the response structure instead.
+
+The custom transfer adapter does not need to check the SHA of the file content
+it has downloaded, git-lfs will do that before moving the final content into
+the LFS store.
+
+##### Progress
+
+In order to support progress reporting while data is uploading / downloading, 
+the transfer process should post messages to stdout as follows before sending
+the final completion message:
+
+```json
+{ "id":"progress", "oid": "22ab5f63670800cc7be06dbed816012b0dc411e774754c7579467d2536a9cf3e", "bytesSoFar": 1234, "bytesSinceLast": 64 }
```

Comment:
Until now, I don't think any of our API JSON has had multi-word keys. Here we're camelcasing them, e.g. `bytesSoFar`, but in the [locking proposal](https://github.com/github/git-lfs/blob/master/docs/proposals/locking_api.md) we've got snakecase going on. Should we standardize on one or the other? I have no opinion of one over the other.